feat: allow creation of initial global environments on first start

Panaetius opened the pull request:
/deploy renku=initial-envs extra-values=dataService.environments.seed=true

leafty reviewed (commented):
Question: is forward compatibility going to be an issue when using an Alembic migration?

What if the schema of environment changes (e.g. add more fields, rename some, remove some)? Is the content of the values file supposed to be using the old schema to make this migration work?

Note: We already have data migrations in `bases/renku_data_services/background_jobs/core.py` and the jobs there can be kept up-to-date with the current schema.

Panaetius commented:
> Question: is forward compatibility going to be an issue when using an Alembic migration?
> 
> What if the schema of environment changes (e.g. add more fields, rename some, remove some)? Is the content of the values file supposed to be using the old schema to make this migration work?
> 
> Note: We already have data migrations in `bases/renku_data_services/background_jobs/core.py` and the jobs there can be kept up-to-date with the current schema.

I on purpose did not depend on the schema in the migration, so it's independent of the orm class. 
My thinking was that if we ever do change the schema, then we should have an accompanying migration that migrates existing envs. So then on a new deployment, this would then create envs with the old schema, and that to be added migration would migrate them the same as any other existing environments.

I intended this to be a one time QoL thing to seed deployments, not a separate way to define environments independent of the admin panel, so changing the value after it ran once wouldn't do anything (though I forgot to mention as much in the values.yaml, will do that). with the main use-case being just having CI deployments be usable more easily (something external contributors struggled with in the past). So as it's a one-time thing in the history of migrations, it only depends on the schema at that point in time/at that point in the migration chain. But yes, that'd mean that the values file would still use the old schema (or we need to e.g. delete this migrations and add a new one if we don't want that behavior). 

We could of course have it be something that runs each time the service starts, that'd also work. At least for us this should not be used for prod anyways, so it shouldn't cause issues, either (e.g. someone deleting the env in the admin panel and it magically reappearing a while later). Migrations seemed the easiest way to have this as a one-time thing in CI deployments.

olevski reviewed (changes requested):
I think we should rewrite this to run at startup when the server starts up. Or as a background job. Which honestly may be better than running at startup.

It is confusing that this will happen only once. So if the admin updates the contents of the environment json spec file. Then the updates will not be reflected. If we make it a background job that run periodically we can just do updates. And most importantly we can import/use the code from other modules and do proper validation.

p.s. I did not realize there was already a review on this. I will read the existing stuff and make edits.

olevski commented:
@Panaetius I have another alternative in mind. Can we make 1 single environment that you can inject like this? So we fully hardcode it in the migration. I think then we do not have to worry about syncing the file contents or even more importantly we do not have to worry about validation and our endpoints breaking because an environment was injected that will fail validation when we serialize.

Then this entry in the list is migrated as with all other environments.

To make my suggestion fully work though we need "latest" tags on the image that we will use for the environment. I dont think the CI pipelines properly assign this. I propose that we pick one image from our renkulab-docker repo and hardcode it. Perhaps jupyterlab or vscode.

Panaetius commented on components/renku_data_services/migrations/versions/450ae3930996_create_initial_global_environments.py:
```diff
@@ -0,0 +1,151 @@
+depends_on = None
+
+
+@dataclass
+class Environment:
+    name: str
+    container_image: str
+    default_url: str
+    port: int = 8888
+    description: str = ""
+    working_directory: str | None = None
+    mount_directory: str | None = None
+    uid: int = 1000
+    gid: int = 1000
+    args: list[str] | None = None
+    command: list[str] | None = None
```
yes if we change the model, then the db won't actually match it at this point in time.

Panaetius commented:
> @Panaetius I have another alternative in mind. Can we make 1 single environment that you can inject like this? So we fully hardcode it in the migration. I think then we do not have to worry about syncing the file contents or even more importantly we do not have to worry about validation and our endpoints breaking because an environment was injected that will fail validation when we serialize.
> 
> Then this entry in the list is migrated as with all other environments.
> 
> To make my suggestion fully work though we need "latest" tags on the image that we will use for the environment. I dont think the CI pipelines properly assign this. I propose that we pick one image from our renkulab-docker repo and hardcode it. Perhaps jupyterlab or vscode.

@olevski oh I like this idea, I've changed it as such. We do have latest tags for jupyter and rstudio, so I've added those. This means someone deploying renku will also get these two environments (though the migration does nothing if environments already exist), but I think that's ok. they can always remove them and having something from the get-go is actually nice imo.
